docs/man/man1/gdal_polygonize.1
.TH "gdal_polygonize" 1 "Mon Nov 20 2017" "GDAL" \" -*- nroff -*-
.ad l
.nh
.SH NAME
gdal_polygonize \- Produces a polygon feature layer from a raster\&.
.SH "SYNOPSIS"
.PP
.PP
.nf
gdal_polygonize.py [-8] [-nomask] [-mask filename] raster_file [-b band]
                [-q] [-f ogr_format] out_file [layer] [fieldname]
.fi
.PP
.SH "DESCRIPTION"
.PP
This utility creates vector polygons for all connected regions of pixels in the raster sharing a common pixel value\&. Each polygon is created with an attribute indicating the pixel value of that polygon\&. A raster mask may also be provided to determine which pixels are eligible for processing\&.
.PP
The utility will create the output vector datasource if it does not already exist, defaulting to GML format\&.
.PP
The utility is based on the GDALPolygonize() function which has additional details on the algorithm\&.
.PP
.IP "\fB\fB-8\fP:\fP" 1c
(GDAL >= 1\&.10) Use 8 connectedness\&. Default is 4 connectedness\&. 
.PP
.IP "\fB\fB-nomask\fP:\fP" 1c
Do not use the default validity mask for the input band (such as nodata, or alpha masks)\&. 
.PP
.IP "\fB\fB-mask\fP \fIfilename\fP:\fP" 1c
Use the first band of the specified file as a validity mask (zero is invalid, non-zero is valid)\&. If not specified, the default validity mask for the input band (such as nodata, or alpha masks) will be used (unless -nomask is specified) 
.PP
.IP "\fB\fIraster_file\fP\fP" 1c
The source raster file from which polygons are derived\&.
.PP
.IP "\fB\fB-b\fP \fIband\fP: \fP" 1c
The band on \fIraster_file\fP to build the polygons from\&. Starting with GDAL 2\&.2, the value can also be set to 'mask', to indicate that the mask band of the first band must be used (or 'mask,band_number' for the mask of a specified band) 
.PP
.IP "\fB\fB-f\fP \fIogr_format\fP\fP" 1c
Select the output format of the file to be created\&. Default is GML\&. 
.PP
.IP "\fB\fIout_file\fP\fP" 1c
The destination vector file to which the polygons will be written\&. 
.PP
.IP "\fB\fIlayer\fP\fP" 1c
The name of the layer created to hold the polygon features\&. 
.PP
.IP "\fB\fIfieldname\fP\fP" 1c
The name of the field to create (defaults to 'DN')\&. 
.PP
.IP "\fB\fB-q\fP:\fP" 1c
The script runs in quiet mode\&. The progress monitor is suppressed and routine messages are not displayed\&. 
.PP
.PP
.PP
.SH "AUTHORS"
.PP
Frank Warmerdam warmerdam@pobox.com 


bindings/python/scripts/gdal_polygonize.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
#******************************************************************************
#  $Id: gdal_polygonize.py 34304 2016-05-27 17:46:17Z rouault $
#
#  Project:  GDAL Python Interface
#  Purpose:  Application for converting raster data to a vector polygon layer.
#  Author:   Frank Warmerdam, warmerdam@pobox.com
#
#******************************************************************************
#  Copyright (c) 2008, Frank Warmerdam
#  Copyright (c) 2009-2013, Even Rouault <even dot rouault at mines-paris dot org>
#
#  Permission is hereby granted, free of charge, to any person obtaining a
#  copy of this software and associated documentation files (the "Software"),
#  to deal in the Software without restriction, including without limitation
#  the rights to use, copy, modify, merge, publish, distribute, sublicense,
#  and/or sell copies of the Software, and to permit persons to whom the
#  Software is furnished to do so, subject to the following conditions:
#
#  The above copyright notice and this permission notice shall be included
#  in all copies or substantial portions of the Software.
#
#  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
#  OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
#  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
#  THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
#  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
#  FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
#  DEALINGS IN THE SOFTWARE.
#******************************************************************************

import sys

from osgeo import gdal
from osgeo import ogr
from osgeo import osr


def Usage():
    print("""
gdal_polygonize [-8] [-nomask] [-mask filename] raster_file [-b band]
                [-q] [-f ogr_format] out_file [layer] [fieldname]
""")
    sys.exit(1)

# =============================================================================
# 	Mainline
# =============================================================================

format = 'GML'
options = []
quiet_flag = 0
src_filename = None
src_band_n = 1

dst_filename = None
dst_layername = None
dst_fieldname = None
dst_field = -1

mask = 'default'

gdal.AllRegister()
argv = gdal.GeneralCmdLineProcessor( sys.argv )
if argv is None:
    sys.exit( 0 )

# Parse command line arguments.
i = 1
while i < len(argv):
    arg = argv[i]

    if arg == '-f':
        i = i + 1
        format = argv[i]

    elif arg == '-q' or arg == '-quiet':
        quiet_flag = 1

    elif arg == '-8':
        options.append('8CONNECTED=8')

    elif arg == '-nomask':
        mask = 'none'

    elif arg == '-mask':
        i = i + 1
        mask = argv[i]

    elif arg == '-b':
        i = i + 1
        src_band_n = int(argv[i])

    elif src_filename is None:
        src_filename = argv[i]

    elif dst_filename is None:
        dst_filename = argv[i]

    elif dst_layername is None:
        dst_layername = argv[i]

    elif dst_fieldname is None:
        dst_fieldname = argv[i]

    else:
        Usage()

    i = i + 1

if src_filename is None or dst_filename is None:
    Usage()

if dst_layername is None:
    dst_layername = 'out'

# =============================================================================
# 	Verify we have next gen bindings with the polygonize method.
# =============================================================================
try:
    gdal.Polygonize
except:
    print('')
    print('gdal.Polygonize() not available.  You are likely using "old gen"')
    print('bindings or an older version of the next gen bindings.')
    print('')
    sys.exit(1)

# =============================================================================
#	Open source file
# =============================================================================

src_ds = gdal.Open( src_filename )

if src_ds is None:
    print('Unable to open %s' % src_filename)
    sys.exit(1)

srcband = src_ds.GetRasterBand(src_band_n)

if mask is 'default':
    maskband = srcband.GetMaskBand()
elif mask is 'none':
    maskband = None
else:
    mask_ds = gdal.Open( mask )
    maskband = mask_ds.GetRasterBand(1)

# =============================================================================
#       Try opening the destination file as an existing file.
# =============================================================================

try:
    gdal.PushErrorHandler( 'CPLQuietErrorHandler' )
    dst_ds = ogr.Open( dst_filename, update=1 )
    gdal.PopErrorHandler()
except:
    dst_ds = None

# =============================================================================
# 	Create output file.
# =============================================================================
if dst_ds is None:
    drv = ogr.GetDriverByName(format)
    if not quiet_flag:
        print('Creating output %s of format %s.' % (dst_filename, format))
    dst_ds = drv.CreateDataSource( dst_filename )

# =============================================================================
#       Find or create destination layer.
# =============================================================================
try:
    dst_layer = dst_ds.GetLayerByName(dst_layername)
except:
    dst_layer = None

if dst_layer is None:

    srs = None
    if src_ds.GetProjectionRef() != '':
        srs = osr.SpatialReference()
        srs.ImportFromWkt( src_ds.GetProjectionRef() )

    dst_layer = dst_ds.CreateLayer(dst_layername, geom_type=ogr.wkbPolygon, srs = srs )

    if dst_fieldname is None:
        dst_fieldname = 'DN'

    fd = ogr.FieldDefn( dst_fieldname, ogr.OFTInteger )
    dst_layer.CreateField( fd )
    dst_field = 0
else:
    if dst_fieldname is not None:
        dst_field = dst_layer.GetLayerDefn().GetFieldIndex(dst_fieldname)
        if dst_field < 0:
            print("Warning: cannot find field '%s' in layer '%s'" % (dst_fieldname, dst_layername))

# =============================================================================
#	Invoke algorithm.
# =============================================================================

if quiet_flag:
    prog_func = None
else:
    prog_func = gdal.TermProgress

result = gdal.Polygonize( srcband, maskband, dst_layer, dst_field, options,
                          callback = prog_func )

srcband = None
src_ds = None
dst_ds = None
mask_ds = None
